data/trie/trieStorageManager.go

```diff
@@ -459,12 +461,32 @@ func (tsm *trieStorageManager) getSnapshotDb(newDb bool) data.DBWriteCacher {
 	}
 
 	if uint32(len(tsm.snapshots)) > tsm.maxSnapshots {
-		tsm.removeSnapshot()
+		if tsm.keepSnapshots == true {
+			tsm.disconnectSnapshot()
+		} else {
+			tsm.removeSnapshot()
+		}
 	}
 
 	return db
 }
 
+func (tsm *trieStorageManager) disconnectSnapshot() {
+	if len(tsm.snapshots) <= 0 {
+		return
+	}
+	snapshot := tsm.snapshots[0]
+	tsm.snapshots = tsm.snapshots[1:]
+
+	if snapshot.IsInUse() {
+		snapshot.MarkForDisconnection()
+		log.Debug("can't disconnect, snapshot is still in use")
+		return
+	}
+
+	disconnectSnapshot(snapshot)
```

Comment:
You should only return the error from disconnectSnapshot() without logging it, and log it here.